### PR TITLE
[mono] Fix constant folding for Math.Round

### LINKED ISSUE #20449: Incorrect rounding for Math.Round(0.5)
## Steps to Reproduce

Use this program:

``` csharp
using System;

public static class Program
{
	public static void Main()
	{
		Console.WriteLine(Math.Round(0.5));
	}
}
```

### Current Behavior

The output is

```
1
```

### Expected Behavior

The expected output is 

```
0
```

The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.math.round?view=netframework-4.8#System_Math_Round_System_Double_) for `Math.Round` says:

> If the fractional component of a is halfway between two integers, one of which is even and the other odd, then the even number is returned.

Mono 5.11.0 and .NET Core return the expected value of 0 from `Math.Round(0.5)`. Mono 6.8.0 returns 1.

## On which platforms did you notice this

[ ] macOS
[ ] Linux
[x] Windows

**Version Used**: 6.8.0

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/mono/mono/issues/20449

We have a logic to fold various Math* operators to constants (https://github.com/mono/mono/pull/9281) and it's slightly incorrect for `Math.Round(double)` which is "round to even".

Btw, we don't need this folding for LLVM since all math functions are LLVM-intrinsics so the folding is done there.

### CODE REVIEW

File: src/libraries/System.Runtime.Extensions/tests/System/Math.cs [hunk 1]
```diff
@@ -2615,5 +2615,67 @@ public static void Round_Decimal_Modes(decimal x, decimal expected, MidpointRoun
             Assert.Equal(expected, Math.Round(x, 0, mode));
             Assert.Equal(expected, decimal.Round(x, 0, mode));
         }
+
+        [Fact]
+        public static void Round_Double_Constant_Arg()
+        {
+            Assert.Equal( 0, Math.Round( 0.5));
+            Assert.Equal( 0, Math.Round( 0.5));
```
Comment (sys27) [inline src/libraries/System.Runtime.Extensions/tests/System/Math.cs:2623]: Same assert as previous one. Is it required?

File: src/mono/mono/mini/intrinsics.c [hunk 2]
```diff
@@ -1763,7 +1763,7 @@ mini_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
 					result = cosh (source);
 					break;
 				case OP_ROUND:
-					result = round (source);
+					result = round (source * 0.5) * 2.0;
```
Comment (lambdageek) [inline src/mono/mono/mini/intrinsics.c:1766]: I'm confused.   Doesn't this give the wrong answer for odd whole numbers? 

if `source = 3.0` then `result = round (3.0 * 0.5) * 2.0` = `round (1.5) * 2.0` = (midpoints round away from zero) = `2.0 * 2.0` = `4.0`

Comment (EgorBo) [reply src/mono/mono/mini/intrinsics.c:1766]: Oops, yeah, copied the wrong fix 🙂

File: src/mono/mono/utils/mono-math.h [hunk 3]
```diff
@@ -80,4 +81,23 @@ inline double mono_trunc (double a)                 { return mono_trunc_double (
...
 #endif
 
+static inline gdouble
+mono_round_to_even (gdouble x)
+{
+	gdouble floor_tmp;
+
+	/* If the number has no fractional part do nothing This shortcut is necessary
+	 * to workaround precision loss in borderline cases on some platforms */
+	if (x == (gdouble)(gint64) x)
```
Comment (lambdageek) [inline src/mono/mono/utils/mono-math.h:91]: you can use `double` and `int64_t` now, don't need the glib.h types

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/mini/intrinsics.c [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 #include <config.h>
 #include <mono/utils/mono-compiler.h>
```
Comment (lambdageek) [inline src/mono/mono/mini/intrinsics.c:6]:
```suggestion
#include <glib.h>
#include <mono/utils/mono-compiler.h>
```

That should fix the C++ build problems in https://github.com/mono/mono/pull/20450

Comment (EgorBo) [reply src/mono/mono/mini/intrinsics.c:7]: Thanks!

**Pull request merged**